lib/caputils.c

```diff
@@ -14,32 +14,40 @@
  * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
 
-#include <stdio.h>
+#include <sys/prctl.h>
+#include <limits.h>
 
 #include "caputils.h"
 #include "pathnames.h"
 
-int cap_last_cap(void)
+static int test_cap(int cap)
 {
-	/* CAP_LAST_CAP is untrustworthy. */
-	static int ret = -1;
-	int matched;
-	FILE *f;
-
-	if (ret != -1)
-		return ret;
-
-	f = fopen(_PATH_PROC_CAPLASTCAP, "r");
-	if (!f) {
-		ret = CAP_LAST_CAP;	/* guess */
-		return ret;
-	}
+	/* prctl returns 0 or 1 for valid caps, -1 otherwise */
+	return prctl(PR_CAPBSET_READ, cap, 0, 0, 0) >= 0;
+}
 
-	matched = fscanf(f, "%d", &ret);
-	fclose(f);
+/* this cache makes cap_last_cap MT-unsafe */
+static int last_cap = 0;
 
-	if (matched != 1)
-		ret = CAP_LAST_CAP;	/* guess */
+int cap_last_cap(void)
+{
+	/* cached last_cap value */
+	if (last_cap) return last_cap;
+
+	/* binary search over capabilities */
+	int cap0 = 0, cap1 = INT_MAX;
+	while (1) {
+		int cap = (cap0 + cap1) / 2;
```

Comment:
This should be an unsigned int, otherwise `cap0 + cap1` can overflow. (This should only happen when the number of capabilities meets or exceeds `INT_MAX/2`.)

If there is some reason not to do this unsigned, then
```
int cap = (cap0 >> 1) + (cap1 >> 1) + ((cap0 & 1) + (cap1 & 1)) >> 1;
```

Reply:
Doing it unsigned sounds great.